doc/Neighbor/Neighbor_Refresh_HLD.md

```diff
@@ -0,0 +1,321 @@
+
+    Key     : Interface + MAC
+    Value   : List of neighbors using this mac
+```
+
+#### 3.3.2.2 Sending ARP/ND Refresh Packets
+
+- For Resolved ARP/ND, Dst MAC will be same as in its NeighCache
+- For Unresolved ARP, Dst MAC will be Broadcast MAC
+- For Unresolved ND, ICMPv6 Echo Request pkt will be sent from application, Kernel will send NS
+- ARP/NS packets will be sent using raw socket 
+  - Separate sockets for ARP and ICMPv6 NS/Echo Request
+- Kernel processing
+  - VLAN tagging & FDB lookup happens based on outgoing interface
+  - On Vlan Interfaces based on the FDB lookup packet will be either flooded or sent to a specific port
+- Soure MAC 
+  - For SAG interfaces, SAG MAC will be used and it will retrived from CONFIG_DB: SAG_GLOBAL
+  - For all other interfaces, Source MAC will be retrived from kernel.
+- Ifname to Index (required for socket send)
+
+#### 3.3.2.3 Refresh Timer
+
+- Traverse the neighbor Cache entries periodically (every 30 secs)
+  - Check refresh timeout has elapsed for every neighbors
+  - If elapsed then send ARP/NS packet
+
+Refresh timeout Calculation:
+
+To avoid sending all ARP/NS packets simultaneously, each neighbor entry will be configured with different refresh timeout value. This refresh timeout value will be based on MAC/ARP/NS aging time.
+
+Refresh timeout will be set whenever the neighbor entry is added/updated in cache, it will also be recomputed after sending the ARP/NS refresh packet.
+
+```
+ARP Reference Timeout  = Min [MAC age, ARP age]
+
+ND Reference Timeout   = Min [MAC age, ND age]
+
+Refresh Timeout = 30% to 70% of [ARP/ND Reference Timeout] 
+```
+
+```
+For example:
+
+    1) MAC Age is less than ARP age
+
+    ARP ageout = 60 mins 
+    MAC ageout = 30 mins
+    Reference Timeout = 30 mins.
+    Refresh timeout will be between 30% to 70% of reference timeout (9 to 21) mins.
+
+    2) ARP Age is less than MAC age
+
+    ARP ageout = 60 mins 
+    MAC ageout = 90 mins
+    Reference Timeout = 60 mins
+    Refresh timeout will be between 30% to 70% of reference timeout (18 to 42) mins.
+```
+
+## 3.4 SyncD
+No Change in SyncD
+
+## 3.5 SAI
+No Change in SAI
+
+## 3.6 CLI
+
+No new CLIs are added as part of this change.
+
+# 4 Flow Diagrams
+
+Neighbor Refresh Cache Flow 
+
+![Neighbor Refresh Cache Diagram](NeighRefreshCacheUpdate.png "Neighbor Refresh Cache Flow")
+
+# 5 Warm Boot Support
+
+Neighbor refresh entries will be added to NEIGH_REFRESH_TABLE during warmrestart state. Therefore the neighbor refresh cache will be created and refresh timeout will become effective after the neighbors are restored.
+
+# 6 Unit Test
+
+1. Verify ARP/ND entry is refreshed periodically before the ARP/ND entry agesout in kernel (or neighbor entries state changes from REACHABLE to STALE)
+2. Verify ARP/ND entry learned on VLAN interfaces is refreshed before their corresponding MAC ageout in HW
+3. Verify ARP/ND entry is refreshed within few seconds if their corresponding MAC in FDB table is deleted.
```

Comment:
Since there is a CoPP limit for ARP/Nbr RX from HW, what if we send more than the allowed Rx limit? how do we handle the retries?

Reply:
Nbr refresh module has a Tx threshold, ideally this tx threshold should be lower than the CoPP Rx limit.